### ISSUE TITLE
[jext] directory listing view

### ISSUE THREAD

**Issue by rgbkrk** [nteract/archived-desktop-app#2439]
Similar to the classic notebook, we need a `/tree` directory listing for the jupyter extension. A component that already uses the contents API is in commuter -- we could put that in core and use them across the jupyter extension and commuter.

**Comment by theengineear** [comment]
## Problem

We need to get to parity with "classic" Jupyter with nteract's jupyter
extension. Right now, we really only have a component in core called
`NotebookApp` that handles the *editing* of a notebook.

## Jupyter Extension requirements

In our Jupyter extension, we can have the following cases:

Note that we don't necessarily know the `type` from the contents api ahead of
time. In this case, we may need to get the file and _then_ redirect.

^^ Is it possible to make a request for *just* the metadata about a file? I.e.,
not get the whole file? Probably not something that needs worrying about yet,
but it'd be wasteful to fetch the whole file and then just redirect.

```
/edit/<PATH> && { type: "notebook" }
/edit/<PATH> && { type: "file" }
/edit/<PATH> && { type: "directory" }

/view/<PATH> && { type: "notebook" }
/view/<PATH> && { type: "file" }
/view/<PATH> && { type: "directory" }

/tree/<PATH> && { type: "notebook" }
/tree/<PATH> && { type: "file" }
/tree/<PATH> && { type: "directory" }
```

## Plan

```
/edit/<PATH> && { type: "notebook" } >> <NotebookApp />
/edit/<PATH> && { type: "file" } >> ??
/edit/<PATH> && { type: "directory" } >> redirect(/tree/<PATH>)

/view/<PATH> && { type: "notebook" } >> redirect(/edit/<PATH>)
/view/<PATH> && { type: "file" } >> ??
/view/<PATH> && { type: "directory" } >> redirect(/tree/<PATH>)

/tree/<PATH> && { type: "notebook" } >> redirect(/view/<PATH>)
/tree/<PATH> && { type: "file" } >> redirect(/view/<PATH>)
/tree/<PATH> && { type: "directory" } >> <Directory />
```

I plan to make a new `Directory` component inside `/core`, though I think we
may want to refactor that later if `/core` ends up too bloated.

I'm not going to handle the `File` view yet, but I wanted to add it into the
mix for discussion.

**Comment by theengineear** [comment]
@rgbkrk ^^ I was just hashing a few things out on my end and I figured I might as well document my thoughts here.

**Comment by theengineear** [comment]
Oop, looks like we also need a plan to add the rest of the `contents` handling.

**Comment by theengineear** [comment]
~~@rgbkrk just to be clear. It looks like the `rx-jupyter` `contents.js` file targets the `notebook` type, right? However, it needs to be generalized to any of the three recognized types?~~ nvm. i think i spoke too soon. It's just the `epics` that seem to only handle `notebook`s right now. I don't need to change anything in the `contents.js` file in `rx-jupter`, correct?

**Comment by rgbkrk** [comment]
Yeah you shouldn't have to change anything for `rx-jupyter`'s `contents` module. The epic is just notebook specific at the moment.

**Comment by rgbkrk** [comment]
> Is it possible to make a request for just the metadata about a file? I.e.,
not get the whole file?

Yes! A `GET` on the contents API allows additional parameters: http://jupyter-api.surge.sh/#!/contents/get_contents_path

`?content=0` will return only the metadata

**Comment by rgbkrk** [comment]
Which I _think_ with `rx-jupyter`'s contents API you'd write:

```
contents.get(serverConfig, path, { content: 0 })
```

We should/could normalize that in rx-jupyter to convert `true` / `false` to `1` and `0` respectively, hiding the quirk.

**Comment by rgbkrk** [comment]
> I plan to make a new Directory component inside /core, though I think we may want to refactor that later if /core ends up too bloated.

Yeah that's totally fine. Commuter would end up using it anyways.

**Comment by rgbkrk** [comment]
Piling on to this issue, since you're setting up other handlers we'll want a path for creating new notebooks. Classic notebook does it like this:

http://localhost:8888/notebooks/Untitled.ipynb?kernel_name=python2

**Comment by rgbkrk** [comment]
Closed by #2680. I'll open new issues for some of the current pitfalls / needs.
